### PR TITLE
Search related tracks in collection

### PR DESCRIPTION
**Pull request by uklotzde**
A simple yet effective extension of the track context menu for exploring and managing the library:

Screenshot from 2020-10-20 00-11-24

Searching for tracks with a similar tempo would be another option.

I have created this feature for personal usage and will not extend it myself, i.e. only provided as a starting point.

### CODE REVIEW

**Comment by uklotzde** [comment]
The CI test failures on Windows are unrelated. @Holzhaus Maybe caused by recent float fixes? These failures seem to become more frequent.

**Comment by uklotzde** [comment]
I must have created a Launchpad issue a long time ago: https://bugs.launchpad.net/mixxx/+bug/1455901

**Comment by uklotzde** [comment]
Update and extended. BPM and year can be added by someone else who needs it.

**Comment by ronso0** [comment]
> Update and extended. BPM and year can be added by someone else who needs it.

So this could be extended to 'fuzzy BPM search' like +- 5 BPM?

**Comment by uklotzde** [comment]
> > Update and extended. BPM and year can be added by someone else who needs it.
> 
> So this could be extended to 'fuzzy BPM search' like +- 5 BPM?

Sure. Just a few lines of code. But I am not keen on discussions about the actual range and configuration options, therefore left out in this PR.

The year is also tricky, because this is a string field. Parsing the string to a date is part of the aoide PR, I already had to do that. Too lazy to backport it.

**Comment by ronso0** [comment]
> > > Update and extended. BPM and year can be added by someone else who needs it.
> > 
> > 
> > So this could be extended to 'fuzzy BPM search' like +- 5 BPM?
> 
> Sure. Just a few lines of code. But I am not keen on discussions about the actual range and configuration options, therefore left out in this PR.

Sure.

> The year is also tricky, because this is a string field. Parsing the string to a date is part of the aoide PR, I already had to do that. Too lazy to backport it.

Could you please share link to the code that does it?

**Comment by uklotzde** [comment]
> > > > Update and extended. BPM and year can be added by someone else who needs it.
> > > 
> > > 
> > > So this could be extended to 'fuzzy BPM search' like +- 5 BPM?
> > 
> > 
> > Sure. Just a few lines of code. But I am not keen on discussions about the actual range and configuration options, therefore left out in this PR.
> 
> Sure.
> 
> > The year is also tricky, because this is a string field. Parsing the string to a date is part of the aoide PR, I already had to do that. Too lazy to backport it.
> 
> Could you please share link to the code that does it?

https://github.com/mixxxdj/mixxx/blob/bdb01601344618b3907c46ab40fb3ac82abe17e5/src/test/aoide_trackexporter_test.cpp#L31

The tests are both the starting point and crucial!!

**Comment by poelzi** [comment]
Lastfm has a query similar api that returns artists with similar style. If we could form a or query, that might be interesting. Or we implement this as some fancy filter for the search query. lastfm_similar("Bakermat") would be replaced with some large artist in ['...', '..', '...', ] query. It seems you can't combine like with in.

**Comment by uklotzde** [comment]
> Lastfm has a query similar api that returns artists with similar style. If we could form a or query, that might be interesting. Or we implement this as some fancy filter for the search query. lastfm_similar("Bakermat") would be replaced with some large artist in ['...', '..', '...', ] query. It seems you can't combine like with in.

Extending the query capabilities is out of scope of this PR.

**Comment by uklotzde** [comment]
If this PR is not reviewed in-time I will delete it and keep it as a commit in my private fork on top of all my other pending changes. The changes introduced here cause a lot of merge conflicts for me and I am not willing to resolve them for an extended period of time.

**Review by Holzhaus — COMMENTED** [review]
Code looks good. I'll give this a try tomorrow.

**Comment by Be-ing** [comment]
I think there's room for a more innovative design that doesn't rely on navigating menus and right clicking (which isn't available on touch screens), but that can be added later as another interface for the same features. For now, adding this to WTrackMenu is an easy and obvious design for a big usability improvement, especially now that WTrackMenu is available in the decks.

File: src/widget/wtrackmenu.h [hunk 1]
```diff
@@ -43,15 +44,17 @@ class WTrackMenu : public QMenu {
...
-                BPM | Color | HideUnhidePurge | FileBrowser | Properties
+                BPM | Color | HideUnhidePurge | FileBrowser | Properties |
+                SearchRelated
     };
     Q_DECLARE_FLAGS(Features, Feature)
 
     WTrackMenu(QWidget* parent,
             UserSettingsPointer pConfig,
-            TrackCollectionManager* pTrackCollectionManager,
+            Library* pLibrary,
```
Comment (Be-ing) [inline src/widget/wtrackmenu.h:57]: Increasing coupling between WTrackMenu and Library smells like not a great design decision to me. Could WTrackMenu instead emit a signal to trigger the query?

File: src/widget/wtrackmenu.cpp [hunk 2]
```diff
@@ -138,6 +139,166 @@ void WTrackMenu::createMenus() {
         //: Reset metadata in right click track context menu in library
         m_pClearMetadataMenu->setTitle(tr("Reset"));
     }
+
+    if (featureIsEnabled(Feature::SearchRelated)) {
+        // TODO: Create a new subclass of QMenu?
```
Comment (Be-ing) [inline src/widget/wtrackmenu.cpp:144]: Yes please. This is IMO too much feature-specific code to cram into createMenus.

File: src/widget/wtrackmenu.h (hunk 1 above)
Comment (Be-ing) [reply src/widget/wtrackmenu.h:57]: Hmm, I'm not sure that would really make a difference though. The parent of WTrackMenu would still need a Library* pointer to make the signal-slot connection. What do you think?

File: src/widget/wtrackmenu.cpp [hunk 3]
```diff
@@ -138,6 +139,166 @@ void WTrackMenu::createMenus() {
...
+                        const auto albumTitleText =
+                                mixxx::escapeTextPropertyWithoutShortcuts(albumTitle);
+                        m_pSearchRelatedMenu->addAction(
+                                tr("Album \"%1\"")
+                                        .arg(albumTitleText),
+                                [this, albumTitle]() {
+                                    m_pLibrary->searchTracksInCollection(
+                                            QStringLiteral("album:\"") +
+                                            albumTitle + QChar('"'));
+                                });
```
Comment (Be-ing) [inline src/widget/wtrackmenu.cpp:270]: This will be really useful for adding a whole album to AutoDJ. :smile:

File: src/widget/wtrackmenu.cpp [hunk 4]
```diff
@@ -153,7 +314,6 @@ void WTrackMenu::createActions() {
     }
 
     if (featureIsEnabled(Feature::LoadTo)) {
-        m_pAddToPreviewDeck = new QAction(tr("Preview Deck"), m_pLoadToMenu);
```
Comment (Be-ing) [inline src/widget/wtrackmenu.cpp:156]: Why was this removed? It seems unrelated to this PR.

File: src/widget/wtrackmenu.cpp [hunk 5]
```diff
@@ -138,6 +139,166 @@ void WTrackMenu::createMenus() {
...
+                                mixxx::escapeTextPropertyWithoutShortcuts(title);
+                        m_pSearchRelatedMenu->addAction(
+                                tr("Title \"%1\"")
+                                        .arg(titleText),
+                                [this, title]() {
+                                    m_pLibrary->searchTracksInCollection(
+                                            QStringLiteral("title:\"") +
+                                            title + QChar('"'));
+                                });
+                    }
```
Comment (Be-ing) [inline src/widget/wtrackmenu.cpp:196]: At first I questioned how this could be useful. Then I realized it could be used to search for remixes of the track. However, this use case isn't very obvious. Any ideas for making it more discoverable? Also, it doesn't work as well for finding similar tracks from a remix unless I delete the `(REMIXER Remix)` suffix manually.

Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:196]: Same here. I noticed how convenient this is when searching for duplicates or cover versions of a track. Manually editing the prepared search query is also an option. Easier than copy&paste a title query manually.

File: src/widget/wtrackmenu.cpp (hunk 3 above)
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:270]: The same applies for the parent folder search, depending on how you organize your music.

File: src/widget/wtrackmenu.cpp (hunk 4 above)
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp:156]: Probably caused by resolving a merge conflict? Damn!

File: src/widget/wtrackmenu.h (hunk 1 above)
Comment (uklotzde) [reply src/widget/wtrackmenu.h:57]: I agree, but since we already need either `Library` or `TrackCollectionManager` passing the top-level `Library` is even easier and cleaner now.

The new WSearchRelatedTracksMenu class will emit a signal that is received by WTrackMenu. The function invocation on `Library` will remain. I am not refactoring the whole design in this PR.

**Comment by uklotzde** [comment]
I need to force push to fix a preceding PR that contains unintended changes caused by conflict resolution while cherry-picking.

**Comment by uklotzde** [comment]
I have also reordered and separated the menu items:
- Musical properties
- Artists
- Anything else

File: src/widget/wtrackmenu.cpp (hunk 4 above)
Comment (uklotzde) [reply src/widget/wtrackmenu.cpp]: Fixed.

**Comment by uklotzde** [comment]
SCons build should be fixed now, just a missing `,`

**Comment by uklotzde** [comment]
This feature turned out to be super helpful for browsing and decluttering my collection. That's why I added more options than initially planned :)

File: src/widget/wtrackmenu.cpp (hunk 2 above)
Comment (Be-ing) [reply src/widget/wtrackmenu.cpp:145]: The TODO comment still needs to be removed.

File: src/widget/wtrackmenu.h (hunk 1 above)
Comment (Be-ing) [reply src/widget/wtrackmenu.h:58]: okay

**Comment by uklotzde** [comment]
@ronso0 I have added the frequently requested context-sensitive bpm range search ;)

File: src/widget/wsearchrelatedtracksmenu.cpp [hunk 6]
```diff
@@ -3,6 +3,20 @@
 #include "track/track.h"
 #include "util/qt.h"
 
+namespace {
+
+constexpr double kRelativeBpmRange = 0.06; // +/-6 %
+
+inline int bpmLowerBound(double bpm) {
```
Comment (Be-ing) [inline src/widget/wsearchrelatedtracksmenu.cpp:10]: I'd prefer to add a direct way to search within a percentage (or absolute) range for numeric search operators than have this calculated in a GUI class. Can you add a TODO comment for that?

**Comment by uklotzde** [comment]
We currently don't have the option to refine an existing search by adding more terms instead of replacing the whole query. This could be useful in conjunction with this context-sensitive search. Selecting between both options on demand is difficult with a context menu.

**Comment by Be-ing** [comment]
> We currently don't have the option to refine an existing search by adding more terms instead of replacing the whole query. This could be useful in conjunction with this context-sensitive search. Selecting between both options on demand is difficult with a context menu.

I was thinking the same.

**Comment by Be-ing** [comment]
I think the main use case for modifying an existing query would be better implemented by extending the queries to be able to reference deck properties (like `bpm:d1~6%` for deck 1's BPM +/- 6%) then creating a concept of a focused/active deck that can be referenced in search queries (like `bpm:df~6%`) which follows the crossfader & volume faders (or manual user deck selection). This way the search query doesn't need to be continually modified.

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 6 above)
Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:10]: I don't think that we need to add an explicit TODO here. It doesn't help. Otherwise, I would need to add TODO to all code saying "this code will probably be reusable, please extract it when needed" ;)

The constants and functions are already separated and extracting this functionality should be straightforward.

Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:10]: Only the query is composed inline to avoid formatting the numbers twice.

Comment (Be-ing) [reply src/widget/wsearchrelatedtracksmenu.cpp:10]: > The constants and functions are already separated and extracting this functionality should be straightforward.

Yes, but my point is this shouldn't be anywhere in wsearchrelatedtracksmenu.cpp.

**Comment by Holzhaus** [comment]
> > > Update and extended. BPM and year can be added by someone else who needs it.
> > 
> > 
> > So this could be extended to 'fuzzy BPM search' like +- 5 BPM?
> 
> Sure. Just a few lines of code. But I am not keen on discussions about the actual range and configuration options, therefore left out in this PR.

Idea: Use the range of tempo fader (e.g. if rate range is 10%, then find all track within that range).

**Comment by Be-ing** [comment]
> Idea: Use the range of tempo fader (e.g. if rate range is 10%, then find all track within that range).

Good idea. Again, this requires search queries to reference decks rather than calculating absolute numbers for a query.

**Comment by uklotzde** [comment]
> > Idea: Use the range of tempo fader (e.g. if rate range is 10%, then find all track within that range).
> 
> Good idea. Again, this requires search queries to reference decks rather than calculating absolute numbers for a query.

Please take into account that the full range in both directions is only available if the current track is played at the nominal tempo, i.e. at the center position! Therefore I used only +-6% which is 3/4 of the classical 8% range. This already results in a wide range.

**Comment by Holzhaus** [comment]
> > > Idea: Use the range of tempo fader (e.g. if rate range is 10%, then find all track within that range).
> > 
> > 
> > Good idea. Again, this requires search queries to reference decks rather than calculating absolute numbers for a query.
> 
> Please take into account that the full range in both directions is only available if the current track is played at the nominal tempo, i.e. at the center position! Therefore I used only +-6% which is 3/4 of the classical 8% range. This already results in a wide range.

Yes. If we later a the concept of a "focused deck", then we could add a way to look up tracks from the deck's track menu using the *current* tempo and therefore find track in the exact range that can be played in sync on the other deck without changing the rate range.

**Comment by uklotzde** [comment]
...and please don't only think in technical terms. If I am explicitly searching for a track with a similar bpm I don't want to get all tracks in a range of +/-10% even if my equipment technically supports this. The number of matching tracks in this range would be huge.

**Comment by uklotzde** [comment]
...and as I predicted the discussion is carried away from reviewing this PR.

**Comment by uklotzde** [comment]
Let's start with this simple MVP and collect feedback about what else is needed instead of getting lost in theoretical discussions.

File: src/widget/wsearchrelatedtracksmenu.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    const auto genre = track.getGenre();
+    if (!genre.isEmpty()) {
+        const auto genreDisplay =
+                mixxx::escapeTextPropertyWithoutShortcuts(genre);
+        addSeparator = addSeparatorBeforeAction(addSeparator);
+        addAction(
+                tr("Genre \"%1\"")
+                        .arg(genreDisplay),
+                [this, genre]() {
```
Comment (Holzhaus) [inline src/widget/wsearchrelatedtracksmenu.cpp:68]: Does this use `connect` internally? Asking for a friend.

**Comment by Be-ing** [comment]
> ...and as I predicted the discussion is carried away from reviewing this PR.
...
Let's start with this simple MVP and collect feedback about what else is needed instead of getting lost in theoretical discussions.

In case I wasn't clear before, these discussions about future improvements shouldn't delay merging this very useful PR. All I'm asking is to add a TODO in the code noting this should be improved further.

**Comment by uklotzde** [comment]
> > ...and as I predicted the discussion is carried away from reviewing this PR.
> > ...
> > Let's start with this simple MVP and collect feedback about what else is needed instead of getting lost in theoretical discussions.
> 
> In case I wasn't clear before, these discussions about future improvements shouldn't delay merging this very useful PR. All I'm asking is to add a TODO in the code noting this should be improved further.

Code can always be improved. TODO comments are only useful if you already know the target state and this information is not yet obvious from the code. I don't have a clear idea where this will end, so I am not able to write such a comment. 

The code also doesn't contain any technical debt, which would be another reason for leaving a TODO or FIXME comment.

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 7 above)
Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:68]: Oh, you are right :see_no_evil: There is an alternative, overloaded (*yikes*) variant that allows to provide the context for the receiver. Since both the action and the menu are part of the same event loop context we can safely omit this redundant context here.

Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:68]: The explicit QObject::connect() is problematic, because you explicitly provide a sender pointer which may live in a different context than the receiver.

**Comment by Be-ing** [comment]
If you really don't want to add a TODO comment, okay. I'm not going to insist on it.

**Comment by uklotzde** [comment]
> If you really don't want to add a TODO comment, okay. I'm not going to insist on it.

It would simply be useless noise if I write something generic without a clear intention or direction.

**Comment by uklotzde** [comment]
@all Please provide feedback on the ordering and formatting of the menu and its items. This is what matters most in the first place.

**Comment by uklotzde** [comment]
Added support for *grouping* and *(release) year*

**Comment by uklotzde** [comment]
Please note that our numeric `year` filter does **not** find any tracks with non-numeric date/time strings! This is a separate issue and cannot be solved here.

**Comment by Holzhaus** [comment]
> @ALL Please provide feedback on the ordering and formatting of the menu and its items. This is what matters most in the first place.

I think this would make sense:

```
Key
BPM
-----
Title
Album
Artist
Album Artist
-----
Year
Genre
Folder
```

The first 2 are probably the most useful and the last 3 are probably the most useful, and I find elements at the top or end of a list easier to find quickly than in the middle of a list.

File: src/widget/wtrackmenu.cpp [hunk 8]
```diff
@@ -957,7 +991,9 @@ void WTrackMenu::addSelectionToPlaylist(int iPlaylistId) {
         return;
     }
 
-    PlaylistDAO& playlistDao = m_pTrackCollectionManager->internalCollection()->getPlaylistDAO();
+    PlaylistDAO& playlistDao = m_pLibrary->trackCollections()
```
Comment (Holzhaus) [inline src/widget/wtrackmenu.cpp:994]:
```suggestion
    const PlaylistDAO& playlistDao = m_pLibrary->trackCollections()
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Unfortunately the key search depends on the notation chosen in the preferences and won't work in all cases. For "Lancelot" it works, but for "Lancelot/Traditional" it doesn't.

**Comment by uklotzde** [comment]
> Unfortunately the key search depends on the notation chosen in the preferences and won't work in all cases. For "Lancelot" it works, but for "Lancelot/Traditional" it doesn't.

The key text parsing is done by `KeyUtils::guessKeyFromText`. I haven't checked if this works as expected, out of scope. The text is provided by the `Track` object and should match the user's preferences.

**Comment by Holzhaus** [comment]
> > Unfortunately the key search depends on the notation chosen in the preferences and won't work in all cases. For "Lancelot" it works, but for "Lancelot/Traditional" it doesn't.
> 
> The key text parsing is done by `KeyUtils::guessKeyFromText`. I haven't checked if this works as expected, out of scope. The text is provided by the `Track` object and should match the user's preferences.
> 
> TL;DR: I will not fix the key search, provided as is.

I think it's just a quoting issue. `~key:7A` works, `~key:7A (Dm)` does not. Adding quotes fixes it: `~key:"7A (Dm)"`.

**Comment by ronso0** [comment]
> > @ALL Please provide feedback on the ordering and formatting of the menu and its items. This is what matters most in the first place.
> 
> The first 2 and the last 3 are probably the most useful, and I find elements at the top or end of a list easier to find quickly than in the middle of a list.

I'd use BPM and Artist the most, and I think searching for Title is used rarely.
So I'd swap the Artist/Album Artist and Title/ Album Title rows:
```
Key
BPM
-----
Artist
Album Artist
Title
Album
-----
Year
Genre
Folder
```

**Comment by uklotzde** [comment]
> > > Unfortunately the key search depends on the notation chosen in the preferences and won't work in all cases. For "Lancelot" it works, but for "Lancelot/Traditional" it doesn't.
> > 
> > 
> > The key text parsing is done by `KeyUtils::guessKeyFromText`. I haven't checked if this works as expected, out of scope. The text is provided by the `Track` object and should match the user's preferences.
> > TL;DR: I will not fix the key search, provided as is.
> 
> I think it's just a quoting issue. `~key:7A` works, `~key:7A (Dm)` does not. Adding quotes fixes it: `~key:"7A (Dm)"`.

Thanks for the hint! I have recently removed the quotes because I thought they were not needed :see_no_evil: Will re-add them.

**Comment by uklotzde** [comment]
> > > Unfortunately the key search depends on the notation chosen in the preferences and won't work in all cases. For "Lancelot" it works, but for "Lancelot/Traditional" it doesn't.
> > 
> > 
> > The key text parsing is done by `KeyUtils::guessKeyFromText`. I haven't checked if this works as expected, out of scope. The text is provided by the `Track` object and should match the user's preferences.
> > TL;DR: I will not fix the key search, provided as is.
> 
> I think it's just a quoting issue. `~key:7A` works, `~key:7A (Dm)` does not. Adding quotes fixes it: `~key:"7A (Dm)"`.

Fixed.

**Comment by uklotzde** [comment]
> > > @ALL Please provide feedback on the ordering and formatting of the menu and its items. This is what matters most in the first place.
> > 
> > 
> > The first 2 and the last 3 are probably the most useful, and I find elements at the top or end of a list easier to find quickly than in the middle of a list.
> 
> I'd use BPM and Artist the most, and I think searching for Title is used rarely.
> So I'd swap the Artist/Album Artist and Title/ Album Title rows:
> 
> ```
> Key
> BPM
> -----
> Artist
> Album Artist
> Title
> Album
> -----
> Year
> Genre
> Folder
> ```

Done.

**Review by Holzhaus — APPROVED** [review]
LGTM, thank you!

File: src/widget/wsearchrelatedtracksmenu.cpp [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+} // namespace
+
+WSearchRelatedTracksMenu::WSearchRelatedTracksMenu(
+        QWidget* parent)
+        : QMenu(tr("Search related Tracks"), parent) {
+}
+
+bool WSearchRelatedTracksMenu::addTriggerSearchAction(
```
Comment (Be-ing) [inline src/widget/wsearchrelatedtracksmenu.cpp:50]:
```suggestion
void WSearchRelatedTracksMenu::addTriggerSearchAction(
```
Why does this return anything? I cannot find a reason. The return value is never used for anything.

File: src/widget/wsearchrelatedtracksmenu.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                    QStringLiteral("~key:\"") +
+                    keyText +
+                    QChar('"');
+            addSeparatorBeforeNextAction = addTriggerSearchAction(
+                    addSeparatorBeforeNextAction,
+                    actionText,
+                    searchQuery);
+        }
+    }
+    {
```
Comment (Be-ing) [inline src/widget/wsearchrelatedtracksmenu.cpp:88]: What is the need for these blocks? It seems that a simple blank line would be adequate to separate what are currently blocks.

File: src/widget/wsearchrelatedtracksmenu.cpp [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+                    addSeparatorBeforeNextAction,
+                    actionText,
+                    searchQuery);
+        }
+    }
+
+    // Title and grouping actions
+    addSeparatorBeforeNextAction = !isEmpty();
+    {
+        const auto title = track.getTitle();
```
Comment (Be-ing) [inline src/widget/wsearchrelatedtracksmenu.cpp:199]: How about searching for trailing text in parentheses and adding another menu entry without that? I think that would work for the remix/original searches in most cases. For example, if the title is "Foo (Bar Remix)", add items to search both `title:"Foo (Bar Remix)"` and `title:"Foo"`

Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:199]: There are too many variations. I leave this (and the following discussion) to a subsequent PR after we collected some feedback.

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 10 above)
Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:88]: To prevent reusing the wrong local variables between menu entries unintentionally.

Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:88]: This could happen easily when copy & pasting code.

Comment (Be-ing) [reply src/widget/wsearchrelatedtracksmenu.cpp:88]: Okay. I wouldn't have written it that way, but it works how it is.

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 11 above)
Comment (Be-ing) [reply src/widget/wsearchrelatedtracksmenu.cpp:199]: okay

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 9 above)
Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:50]: It is used if you look closely. This allows to write the invocation and resetting of the `addSeparatorBeforeNextAction` in a single line of code.

Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:50]: I will pass it as a pointer instead (in/out param)

Comment (Be-ing) [reply src/widget/wsearchrelatedtracksmenu.cpp:50]: Oh, I see that now. There's likely a more straightforward way to write this, but it's not immediately jumping out to me so I'm fine leaving it how it is.

Comment (Be-ing) [reply src/widget/wsearchrelatedtracksmenu.cpp:50]: > I will pass it as a pointer instead (in/out param)

That seems like a better way.

File: src/widget/wsearchrelatedtracksmenu.cpp (hunk 11 above)
Comment (uklotzde) [reply src/widget/wsearchrelatedtracksmenu.cpp:199]: Feel free to start working on a `class TitleParser` that will be needed for this purpose ;)

**Comment by Be-ing** [comment]
Thanks for working on this!

**Comment by ronso0** [comment]
:+1: in conjunction with Remove from Disk #3212 this is a killer tool for cleaning up my library!

only the location string shown in the `Folder: `: is that needed? it expands the menu a lot.

**Comment by uklotzde** [comment]
> +1 in conjunction with Remove from Disk #3212 this is a killer tool for cleaning up my library!
> 
> only the location string shown in the `Folder: `: is that needed? it expands the menu a lot.

That's exactly the reason why we need an MVP of every new feature for collecting early feedback instead of getting lost in detailed discussions about potential edge cases ;) Work continues in #3215

**Comment by ronso0** [comment]
for everyone who like me has to lookup MVP: minimum viable product
